src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs

```diff
@@ -340,14 +407,49 @@ private void Write(RuntimeModel.IObjectWriter writer, bool hashing, Action<Packa
             foreach (var pair in _projects)
             {
                 var project = pair.Value;
+                WriteProject(writer, hashing, writeAction, project, projectNameToHashCode);
+            }
+
+            writer.WriteObjectEnd();
+            writer.WriteObjectEnd();
+        }
+
+        private void WriteProject(IObjectWriter writer, bool hashing, Action<PackageSpec, IObjectWriter, bool, IEnvironmentVariableReader> writeAction, PackageSpec project, Dictionary<string, string>? projectNameToHashCode)
+        {
+            if (hashing && projectNameToHashCode != null)
+            {
+                string? projectHash = null;
+
+                lock (projectNameToHashCode)
```

Comment:
For large restores, I wonder if it would be better to use a per-project lock instead of locking the entire dictionary like we have in TaskResultCache?

Reply:
A per project lock would involve some sort of a lock lookup which itself would have the same issue. 
Looking up is going to be really fast, so I don't expect much contention here.

Reply:
The lock is held for a very short time, because all it's doing is a lookup, but if multiple threads concurrently have a cache miss, then they're all going to take hashes of the object at the same time.

Is hashing zero alloc, or low alloc?  If not, blocking the thread might be less impactful than having multiple threads do duplicate work with allocations that might cause GC, since GC will block all threads.

Reply:
I don't want hashing 1 project to stop restore from hashing other projects, so I'd need per project locks. 
I think that would theoretically be lower allocations, but with the newer hashing algorithm they're already significantly lower, so the biggest benefit in the testing is time.

I can try a more complex implementation if you feel too strongly, but I think this is one of those where I don't perfect to be the enemy of really really good.